### ISSUE TITLE
High  Bitrate Music

### ISSUE THREAD

**Issue by Bigsteve999** [koying/SPMC#150]
Hi, any plans to allow high bitrate audio over pass through? Using the Nvidia Shield to play a 192khz flac only allows 48khz in SPMC but if I play it in Google music its at the full 192khz.

I know my stupid ear can only hear 20khz at best but I can't help but feel I'm missing out.

**Comment by cesarious** [comment]
+1 on HD flacs

also would be nice to have support to dts audio multichannel encoded files.

**Comment by koying** [comment]
Be sure it will be implemented as soon as technically possible

**Comment by mremulator** [comment]
Any news on 24-bit 192Khz playback through SPMC? Still 48Khz through my Shield/AV receiver. Fine through ES File Explorer & Google Music Player, which is far from ideal.

Quick question... Would an external USB DAC work, or would it still be re sampled by the audiotrack API?

**Comment by LadislavParizek** [comment]
Please, where is technical problem? Audio passthrough for Video files excelent works, even TrueHD, and DTS HD MA with big bitrate.
This is my test of HD audio files
- FLAC 96/24 5.1 --> in SPMC is transcoded to PCM 48 kHz 5.1, in Plex same.
- FLAC 96/24 2.0 --> in SPMC is transcoded to PCM 48 kHz 2.0, in Plex PCM 96/24 2.0
- FLAC 192/24 2.0 --> in SPMC is transcoded to PCM 48 kHz 2.0, in Plex PCM 192/24 2.0

On my 4 years old Xstreamer Prodigy all works without problem.
Please Add New radiobutton for Audio passthrough for Audio files. I believe there is no HW problem on Nvidia Shield Android TV.

**Comment by mremulator** [comment]
Just to say my USB DAC (DacMagic 100) works a treat. Upsamples everything to 24/192, but works.

**Comment by jeroen-es** [comment]
Any status update on this? I found this (rather old) topic: http://forum.kodi.tv/showthread.php?tid=259408 which seems to describe the source of the problem. But no solution so far so it seems..

**Comment by smoothquark** [comment]
Hi! I am waiting for this as well.

**Comment by koying** [comment]
Plex, Google Music ar al are using the builtin, high-level, MediaPlayer, which seems to support 24/96.
Kodi, Exoplayer et al are using the MediaCodec/AudioTrack interface, which doesn't expose 24bits audio.

That didn't change in N (http://androidxref.com/7.0.0_r1/xref/frameworks/av/media/libmedia/AudioTrack.cpp#1328) so just forget about it in Kodi/SPMC/MrMC

**Comment by jeroen-es** [comment]
So isn't there an option to use that builtin mediaplayer in kodi/spmc, at least for these types of files? If Plex can do it I suppose it should be possible for kodi as well. Or am I missing something here?

**Comment by smoothquark** [comment]
The audio system in android leaves a lot to be desired, unfortunately. Even though the Nvidia Shield Android TV hardware is great and quite powerful, it is quite limited by the android implementation. Even Plex and Google Music isn't great - 16/44.1 gets converted to 16/48 and cannot play multi-channel 24/96 or 24/192. I guess as ALSA is not exposed, we are rather limited.

I tried compiling mpd to play via my USB DAC, and ran into the same problem with SLES. I was hoping koying may be able to weave some magic, but I think the audio implementation within android is the fundamental problem. I cannot see that being changed as it will break a lot of current software? 

UAPP (USB Audio Player Pro) (http://www.extreamsd.com/index.php/products/usb-audio-player-pro) seems to provide their own proprietary solution via a USB DAC but not HDMI. Does not run on Android TV though. I guess the only possible solution will be to forego android and run LibreELEC on the Nivida Shield Android TV, if it ever crops up. Then, there would be other issues.

I guess the only possible solution is to have a RPi running mpd, connected to a USB DAC and using the mpd add-on in Kodi/SPMS. There is a problem though - How to have HDMI video on one input and audio in another? Some AVRs can achieve this easily, but not so easily in mine. Moreover, would have to rely on Kodi/SPMC decoding DTS-HD and DolbyHD tracks and pass via the USB DAC.

So basically, I think we are stuffed!

**Comment by Bigsteve999** [comment]
Hmmm. Would argue that while it would be nice to have and given that you can only really hear at around 20khz anyway that perhaps this is not the most important thing in the world. Certainly not enough to warrant the risk of messing everything else. Unless of course Koying says that it would be any different.

**Comment by koying** [comment]
> Unless of course Koying says that it would be any different

Nope ;)

Although, in all honesty, you're confused between the frequency range of the human ear and the sampling frequency of digital audio; they are unrelated.

I can understand the quality difference (although probably unnoticeable for most of us) between 16 and 24 bits, but I'm actually not sure where is the quality upscale between a 48khz and 192khz sampling rate...

**Comment by MrMC** [comment]
OpenSL ES or native audiotrack might bypass these limitations.

**Comment by koying** [comment]
> native audiotrack

Stagefright style? Tempting, but linking to system libs is explicitely denied as of API 24, so no future...

**Comment by smoothquark** [comment]
Just had a thought... Is it possible to have a setting for flac whereby it is wrapped as say DTS-HD? Should work for 24/96 and 24/192 both stereo and multichannel? DTS-HD can then be bitstreamed. This should bypass the audio track limitation? How about 44.1 and 88.2? Can this also be used to bitstream DSD?

**Comment by natetg57** [comment]
I really appreciate all of the work that @koying and other developers have done with Kodi and SPMC. I've been really happy with my NVidia Shield for the past year. But I compared some 24/96 tracks played through my laptop with HDMI to the same track played on the Shield and unfortunately the laptop sounded slightly better. I also have highly resolving speakers with ribbon tweeters. My question is, Is it realistic to think the Shield will ever be able to play multichannel FLACs at full resolution? If not, I'll look into getting another device. Thanks so much for your help!
